README.md
<div>
    <!-- Travis badge -->
    <a href="https://travis-ci.com/kiwiroy/mojo-file-role-ingest">
      <img src="https://travis-ci.com/kiwiroy/mojo-file-role-ingest.svg?branch=master"
           alt="Travis Build Status" />
    </a>
    <!-- Kritika badge -->
    <a href="https://kritika.io/users/kiwiroy/repos/9231669397817641/heads/master/">
      <img src="https://kritika.io/users/kiwiroy/repos/9231669397817641/heads/master/status.svg?type=score%2Bcoverage%2Bdeps"
           alt="Kritika Analysis Status" />
    </a>
    <!-- Coveralls badge -->
    <a href="https://coveralls.io/github/kiwiroy/mojo-file-role-ingest?branch=master">
      <img src="https://coveralls.io/repos/github/kiwiroy/mojo-file-role-ingest/badge.svg?branch=master"
           alt="Coverage Status" />
    </a>
</div>

# NAME

Mojo::File::Role::Ingest - Add ingest method to a file.

# SYNOPSIS

    $file = path('file.txt')->with_roles('+Ingest');
    # Mojo::Collection
    $lines = $file->ingest('Mojo::File::Role::Lines');

    @lines;
    $file->ingest('+Lines', {}, sub { push @lines, $_ unless m/^#/ });

    $ext_to_role = {csv => '+CSV', txt => '+Lines'};
    $file->list->each(sub {
      $_->ingest()
    })

# DESCRIPTION

Ingesting the contents of a [Mojo::File](https://metacpan.org/pod/Mojo%3A%3AFile) is a common activity and can be
achieved using ["slurp" in Mojo::File](https://metacpan.org/pod/Mojo%3A%3AFile#slurp) to read all the data into memory. Parsing
the data in the file is also a common task and Mojo::File::Role::Ingest provides
a useful syntax to delegate to a [role](https://metacpan.org/pod/Role%3A%3ATiny) that provides a `parse`
method and have it appropriately consume the file's contents.

A good model is to either return a [Mojo::Collection](https://metacpan.org/pod/Mojo%3A%3ACollection) of records from the file
or on a record by record basis call a supplied callback.

# METHODS

Mojo::File::Role::Ingest composes the following method. The examples below use
the included [Mojo::File::Role::Lines](https://metacpan.org/pod/Mojo%3A%3AFile%3A%3ARole%3A%3ALines).

## ingest

    # Mojo::Collection of lines
    $lines = $file->ingest('+Lines');
    
    # same, but change the end of line ($INPUT_RECORD_SEPARATOR)
    $lines = $file->ingest('+Lines', {eol => "//\n"});
    
    # use a callback to collect lines
    @lines;
    $file->ingest('+Lines', {}, sub { push @lines, $_ unless m/^#/ });
    
    # generally
    $file->ingest($role, $options, $cb);

["ingest"](#ingest) delegates to the supplied [role's](https://metacpan.org/pod/Role%3A%3ATiny) `parse` method. The
`parse` method will be passed the `$options` and `$cb`.

The `$options` hashref will contain the following keys, as a minimum, when
["parse"](#parse) is called by ["ingest" in Mojo::File::Role::Ingest](https://metacpan.org/pod/Mojo%3A%3AFile%3A%3ARole%3A%3AIngest#ingest).

- _encoding_

    The encoding to ["decode" in Mojo::Util](https://metacpan.org/pod/Mojo%3A%3AUtil#decode) with defaulting to `UTF-8`.

- _eol_

    The end of line string defaulting to `$/`

- _is\_large_

    A Boolean flag to notify if the file has a byte size larger than
    `MOJO_MAX_MEMORY_SIZE`.

- _mode_

    The mode to open the file with in the case a callback is used and defaulting to
    `<:encoding(UTF-8)`.

# SEE ALSO

- [Mojo::File](https://metacpan.org/pod/Mojo%3A%3AFile)
- [Mojo::File::Role::IngestI](https://metacpan.org/pod/Mojo%3A%3AFile%3A%3ARole%3A%3AIngestI)
- [Mojo::File::Role::Lines](https://metacpan.org/pod/Mojo%3A%3AFile%3A%3ARole%3A%3ALines)
- [mojolicious/mojo#1392](https://github.com/mojolicious/mojo/issues/1392)
- [mojolicious/mojo#1478](https://github.com/mojolicious/mojo/issues/1478)


Changes
Revision history for perl distribution Mojo-File-Role-Ingest

0.01 Not Released
 - Started project


Makefile.PL
# Generated by git-ship. See 'git-ship --man' for help or https://github.com/jhthorsen/app-git-ship
use utf8;
use ExtUtils::MakeMaker;
my %WriteMakefileArgs = (
  NAME           => 'Mojo::File::Role::Ingest',
  AUTHOR         => 'Roy Storey <kiwiroy@users.noreply.github.com>',
  LICENSE        => 'artistic_2',
  ABSTRACT_FROM  => 'lib/Mojo/File/Role/Ingest.pm',
  VERSION_FROM   => 'lib/Mojo/File/Role/Ingest.pm',
  EXE_FILES      => [qw()],
  OBJECT         => '',
  BUILD_REQUIRES => {}
,
  TEST_REQUIRES  => {
  'Test::More' => '0'
}
,
  PREREQ_PM      => {
  'Mojolicious' => '8.23',
  'Role::Tiny' => '2.000001',
  'perl' => 'v5.20.0'
}
,
  META_MERGE     => {
    'dynamic_config' => 0,
    'meta-spec'      => {version => 2},
    'resources'      => {
      bugtracker => {web => 'https://github.com/kiwiroy/mojo-file-role-ingest/issues'},
      homepage   => 'https://github.com/kiwiroy/mojo-file-role-ingest',
      repository => {
        type => 'git',
        url  => 'https://github.com/kiwiroy/mojo-file-role-ingest.git',
        web  => 'https://github.com/kiwiroy/mojo-file-role-ingest',
      },
    },
    'x_contributors' => []
,
  },
  test => {TESTS => (-e 'META.yml' ? 't/*.t' : 't/*.t xt/*.t')},
);

unless (eval { ExtUtils::MakeMaker->VERSION('6.63_03') }) {
  my $test_requires = delete $WriteMakefileArgs{TEST_REQUIRES};
  @{$WriteMakefileArgs{PREREQ_PM}}{keys %$test_requires} = values %$test_requires;
}

WriteMakefile(%WriteMakefileArgs);


lib/Mojo/File/Role/Lines.pm
package Mojo::File::Role::Lines;
use Mojo::Base -role, -signatures;
use Mojo::Collection 'c';
use Mojo::Util 'decode';

our $VERSION = '0.01';

sub parse ($self, $opts, $cb = undef) {
  return _readlines($self->open($opts->{mode}), $opts, $cb) if $cb;
  return if $opts->{is_large};
  return c(split $opts->{eol} => decode $opts->{encoding} => $self->slurp);
}

sub _readlines ($fh, $opts, $cb) {
  local $/ = $opts->{eol};
  while (readline $fh) {
    chomp;
    $cb->(decode $opts->{encoding}, $_);
  }
}

1;

=encoding utf8

=head1 NAME

Mojo::File::Role::Lines - Ingest a file as a collection of lines.

=head1 SYNOPSIS

  $file = path('file.txt')->with_roles('+Ingest');
  # Mojo::Collection
  $lines = $file->ingest('+Lines');

  @lines;
  $file->ingest('+Lines', {}, sub { push @lines, $_ unless m/^#/ });

=head1 DESCRIPTION

Ingest a file as a collection of lines or collect lines with a callback.

=head1 METHODS

L<Mojo::File::Role::Lines> implements L<Mojo::File::Role::IngestI> and as such
composes the following methods.

=head2 parse

The L</"parse"> method will either parse the file contents into a
L<Mojo::Collection> of lines, or call the supplied callback once per line.

=over 4

=item encoding

=item eol

=item mode

=back

=cut


lib/Mojo/File/Role/IngestI.pm
package Mojo::File::Role::IngestI;

use Mojo::Base -role;

requires 'parse';

1;

=encoding utf8

=head1 NAME

Mojo::File::Role::IngestI - Interface for Ingest

=head1 SYNOPSIS

  # compose Mojo::File::Role::Ingest
  my $obj = Mojo::File->new->with_roles('+Ingest');
  $obj->ingest('+Lines');

=head1 DESCRIPTION

A simple role to ensure composition of Ingest compatible roles have c<parse>.

=head1 METHODS

Mojo::File::Role::IngestI B<requires> the following methods as they will be
called by L<Mojo::File::Role::Ingest/"ingest">.

=head2 parse

  $file->parse($options, $callback);

L<Mojo::File::Role::Ingest/"ingest"> will call L</"parse"> with a hashref
C<$options> and possibly a C<$callback>. If no C<$callback> is supplied the
method is expected to return a L<Mojo::Collection> of I<"things"> that are
defined for the file format being ingested. If a C<$callback> is supplied it is
expected to be called for each record entry in the file being ingested.

The C<$options> hashref will contain the following keys when L</"parse"> is
called by L<Mojo::File::Role::Ingest/"ingest">.

=over 4

=item I<encoding>

The encoding to L<Mojo::Util/"decode"> with defaulting to C<UTF-8>.

=item I<eol>

The end of line string defaulting to C<$/>

=item I<is_large>

A Boolean flag to notify if the file has a byte size larger than
C<MOJO_MAX_MEMORY_SIZE>.

=item I<mode>

The mode to open the file with in the case a callback is used and defaulting to
C<<< <:encoding(UTF-8) >>>.

=back

=cut


lib/Mojo/File/Role/Ingest.pm
package Mojo::File::Role::Ingest;

use Mojo::Base -role, -signatures;
use Mojo::File 'path';
use constant MAX_MEMORY => $ENV{MOJO_MAX_MEMORY_SIZE} || 262144;

our $VERSION = '0.01';

sub ingest {
  my ($self, $class, $opts) = (shift, shift, shift || {});
  return unless defined $class;
  return unless -r $self && !-d _;
  return if $class =~ /[+:]IngestI?$/;
  return
    unless my $composed
    = eval { path(${$self})->with_roles($class, '+IngestI') };
  $opts->{is_large} = !!((-s ${$self}) > MAX_MEMORY);
  $opts->{eol}      //= $/;
  $opts->{encoding} //= 'UTF-8';
  # uncoverable condition false count:1
  # uncoverable condition false count:2
  $opts->{mode}     //= sprintf '<:encoding(%s)' => $opts->{encoding};
  return $composed->parse($opts, @_);
}

1;

=encoding utf8

=begin html

<!-- Travis badge -->
<a href="https://travis-ci.com/kiwiroy/mojo-file-role-ingest">
  <img src="https://travis-ci.com/kiwiroy/mojo-file-role-ingest.svg?branch=master"
       alt="Travis Build Status" />
</a>
<!-- Kritika badge -->
<a href="https://kritika.io/users/kiwiroy/repos/9231669397817641/heads/master/">
  <img src="https://kritika.io/users/kiwiroy/repos/9231669397817641/heads/master/status.svg?type=score%2Bcoverage%2Bdeps"
       alt="Kritika Analysis Status" />
</a>
<!-- Coveralls badge -->
<a href="https://coveralls.io/github/kiwiroy/mojo-file-role-ingest?branch=master">
  <img src="https://coveralls.io/repos/github/kiwiroy/mojo-file-role-ingest/badge.svg?branch=master"
       alt="Coverage Status" />
</a>

=end html

=head1 NAME

Mojo::File::Role::Ingest - Add ingest method to a file.

=head1 SYNOPSIS

  $file = path('file.txt')->with_roles('+Ingest');
  # Mojo::Collection
  $lines = $file->ingest('Mojo::File::Role::Lines');

  @lines;
  $file->ingest('+Lines', {}, sub { push @lines, $_ unless m/^#/ });

  $ext_to_role = {csv => '+CSV', txt => '+Lines'};
  $file->list->each(sub {
    $_->ingest()
  })

=head1 DESCRIPTION

Ingesting the contents of a L<Mojo::File> is a common activity and can be
achieved using L<Mojo::File/"slurp"> to read all the data into memory. Parsing
the data in the file is also a common task and Mojo::File::Role::Ingest provides
a useful syntax to delegate to a L<role|Role::Tiny> that provides a C<parse>
method and have it appropriately consume the file's contents.

A good model is to either return a L<Mojo::Collection> of records from the file
or on a record by record basis call a supplied callback.

=head1 METHODS

Mojo::File::Role::Ingest composes the following method. The examples below use
the included L<Mojo::File::Role::Lines>.

=head2 ingest

  # Mojo::Collection of lines
  $lines = $file->ingest('+Lines');
  
  # same, but change the end of line ($INPUT_RECORD_SEPARATOR)
  $lines = $file->ingest('+Lines', {eol => "//\n"});
  
  # use a callback to collect lines
  @lines;
  $file->ingest('+Lines', {}, sub { push @lines, $_ unless m/^#/ });
  
  # generally
  $file->ingest($role, $options, $cb);

L</"ingest"> delegates to the supplied L<role's|Role::Tiny> C<parse> method. The
C<parse> method will be passed the C<$options> and C<$cb>.

The C<$options> hashref will contain the following keys, as a minimum, when
L</"parse"> is called by L<Mojo::File::Role::Ingest/"ingest">.

=over 4

=item I<encoding>

The encoding to L<Mojo::Util/"decode"> with defaulting to C<UTF-8>.

=item I<eol>

The end of line string defaulting to C<$/>

=item I<is_large>

A Boolean flag to notify if the file has a byte size larger than
C<MOJO_MAX_MEMORY_SIZE>.

=item I<mode>

The mode to open the file with in the case a callback is used and defaulting to
C<<< <:encoding(UTF-8) >>>.

=back

=head1 SEE ALSO

=over 4

=item L<Mojo::File>

=item L<Mojo::File::Role::IngestI>

=item L<Mojo::File::Role::Lines>

=item L<mojolicious/mojo#1392|https://github.com/mojolicious/mojo/issues/1392>

=item L<mojolicious/mojo#1478|https://github.com/mojolicious/mojo/issues/1478>

=back

=cut
